src/Json/Value.php

```diff
@@ -635,37 +636,80 @@ public function ¿timestamp(): ?DateTimeImmutable
+     *  string- and int-backed enums. Also supports enums from the package `vcn/enum`.
      *
-     * @template T of Enum
+     *  For string- and int-backed enums, the JSON value should equal the type and value of a value backing an enum case.
+     *
+     *  For enums from package `vcn/enum`, the value should be a string and should equal the name of the enum case.
+     *
+     * @template T of BackedEnum|Enum
      *
      * @phpstan-param class-string<T> $className
      *
      * @phpstan-return T
      *
-     * @throws Exception\AssertionFailed If this value is not a string, or it is not any of the Enum names from
-     *                                   $className.
-     * @throws Exception\Runtime         If $className does not exist, or does not extend Enum.
+     * @throws Exception\AssertionFailed If the provided value could not be interpreted as a case of the provided
+     *                                    enum class.
+     * @throws Exception\Runtime         If $className does not exist, or is not a valid enum.
      */
-    public function enum(string $className): Enum
+    public function enum(string $className): Enum | BackedEnum
     {
+        if (!class_exists($className)) {
+            throw new Exception\Runtime(sprintf("Class %s does not exist.", $className));
+        }
+
+        // support native enums
+        if (is_subclass_of($className, BackedEnum::class)) {
+            $value = $this->value;
+
+            if ($className::cases() === []) {
+                throw new Exception\AssertionFailed(
+                    sprintf(
+                        "Expected field %s to be enumeration constant '%s', but the enumeration itself is empty.",
+                        $this->pointer,
+                        $value
+                    )
+                );
+            }
+
+            try {
+                return $className::from($value);
+            } catch (ValueError) {
+                $failedAssertions = array_map(
+                    function (BackedEnum $enum) use ($value) {
+                        return new Exception\AssertionFailed(
+                            sprintf(
+                                "Expected %s to be enumeration constant '%s', '%s' given.",
+                                $this->getPointer(),
+                                $enum->value,
+                                $value
+                            )
+                        );
+                    },
+                    $className::cases()
+                );
+
+                throw Json\Exception\ManyAssertionsFailed::fromFailedAssertions(
+                    $failedAssertions[0],
+                    ...array_slice($failedAssertions, 1)
+                );
+            }
+        }
+
+        // BackedEnum has been handled. Additional valid cases must be a vcn\lib\enum, because regular UnitEnum is not supported
         if (!class_exists(Enum::class)) {
-            // @codeCoverageIgnoreStart
```

Comment:
I'd prefer having 100% coverage, explicitly annotating lines not covered.

It makes those lines stand out, a conscious omission, a heads up for future readers.

Reply:
reverted this change.